src/integration-tests/socks-echo/integration.ts

```diff
@@ -0,0 +1,168 @@
+    var authRequest = Socks.composeAuthHandshakeBuffer([Socks.Auth.NOAUTH]);
+    connection.send(authRequest);
+    return connection.receiveNext().then((buffer:ArrayBuffer) : Promise<ArrayBuffer> => {
+      var auth = Socks.interpretAuthResponse(buffer);
+      if (auth != Socks.Auth.NOAUTH) {
+        throw new Error('Unexpected auth value: ' + auth);
+      }
+
+      var request :Socks.Request = {
+        version: Socks.VERSION5,
+        command: Socks.Command.TCP_CONNECT,
+        destination: {
+          addressType: Socks.AddressType.IP_V4,
+          endpoint: webEndpoint,
+          addressByteLength: 7
+        }
+      };
+      connection.send(Socks.composeRequestBuffer(request));
+      return connection.receiveNext();
+    }).then((buffer:ArrayBuffer) : Tcp.Connection => {
+      var responseEndpoint = Socks.interpretRequestResponse(buffer);
+      if (responseEndpoint.address != webEndpoint.address) {
+        throw new Error('unexpected address: ' + responseEndpoint.address);
+      }
+      if (responseEndpoint.port != webEndpoint.port) {
+        throw new Error('response indicates incorrect port: ' +
+                        responseEndpoint.port);
+      }
+      return connection;
+    });
+  }
+
+  public singleEchoTest = (contents:ArrayBuffer) : Promise<ArrayBuffer> => {
+    try {
+      return Promise.all([this.startSocksPair_(), this.startEchoServer_()])
+          .then((endpoints:Net.Endpoint[]) : Promise<Tcp.Connection> => {
+            var socksEndpoint = endpoints[0];
+            var echoEndpoint = endpoints[1];
+            return this.connectThroughSocks_(socksEndpoint, echoEndpoint);
+          }).then((connection:Tcp.Connection) => {
+            connection.send(contents);
+            return connection.receiveNext();
+          });
+    } catch (e) {
+      return Promise.reject(e.message + ' ' + e.stack);
+    }
+  }
+
+  public parallelEchoTest = (contents:ArrayBuffer[]) : Promise<ArrayBuffer[]> => {
+    try {
+      return Promise.all([this.startSocksPair_(), this.startEchoServer_()])
+          .then((endpoints:Net.Endpoint[]) : Promise<Tcp.Connection> => {
+            var socksEndpoint = endpoints[0];
+            var echoEndpoint = endpoints[1];
+            return this.connectThroughSocks_(socksEndpoint, echoEndpoint);
+          }).then((connection:Tcp.Connection) => {
+            contents.forEach(connection.send);
+            var received :ArrayBuffer[] = [];
+            return new Promise<ArrayBuffer[]>((F, R) => {
+              connection.dataFromSocketQueue.setSyncHandler((echo:ArrayBuffer) => {
+                received.push(echo);
+                if (received.length == contents.length) {
+                  F(received);
+                }
+              });
+            });
+          });
+    } catch (e) {
+      return Promise.reject(e.message + ' ' + e.stack);
+    }
+  }
+
+  public parallelEchoTest = (contents:ArrayBuffer[]) : Promise<ArrayBuffer[]> => {
```

Comment:
Resolved.  (parallelEchoTest was defined twice here, but there was a matching typo in the jasmine tests that call parallelEchoTest in both the "parallel" and "serial" tests.  I've fixed both sides and confirmed that the tests still pass.)